test/integration/h2_fuzz.cc

```diff
@@ -0,0 +1,249 @@
+    const uint32_t stream_idx = proto_frame.malformed_request_with_zerolen_header().stream_index();
+    const absl::string_view host = proto_frame.malformed_request_with_zerolen_header().host();
+    const absl::string_view path = proto_frame.malformed_request_with_zerolen_header().path();
+    ENVOY_LOG_MISC(trace, "Sending malformed_request_with_zerolen_header");
+    h2_frame = Http2Frame::makeMalformedRequestWithZerolenHeader(stream_idx, host, path);
+    break;
+  }
+  case test::integration::H2TestFrame::kRequest: {
+    const uint32_t stream_idx = proto_frame.request().stream_index();
+    const absl::string_view host = proto_frame.request().host();
+    const absl::string_view path = proto_frame.request().path();
+    ENVOY_LOG_MISC(trace, "Sending request");
+    h2_frame = Http2Frame::makeRequest(stream_idx, host, path);
+    break;
+  }
+  case test::integration::H2TestFrame::kPostRequest: {
+    const uint32_t stream_idx = proto_frame.post_request().stream_index();
+    const absl::string_view host = proto_frame.post_request().host();
+    const absl::string_view path = proto_frame.post_request().path();
+    ENVOY_LOG_MISC(trace, "Sending post request");
+    h2_frame = Http2Frame::makePostRequest(stream_idx, host, path);
+    break;
+  }
+  case test::integration::H2TestFrame::kGeneric: {
+    const absl::string_view frame_bytes = proto_frame.generic().frame_bytes();
+    ENVOY_LOG_MISC(trace, "Sending generic frame");
+    h2_frame = Http2Frame::makeMalformedFrame(frame_bytes);
+    break;
+  }
+  default:
+    ENVOY_LOG_MISC(debug, "Proto-frame not supported!");
+    break;
+  }
+
+  write_func(h2_frame);
+}
+
+void H2FuzzIntegrationTest::replay(const test::integration::H2CaptureFuzzTestCase& input,
+                                   bool ignore_response) {
+  PERSISTENT_FUZZ_VAR bool initialized = [this]() -> bool {
+    initialize();
+    fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
+    return true;
+  }();
+  UNREFERENCED_PARAMETER(initialized);
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("http"));
+  FakeRawConnectionPtr fake_upstream_connection;
+  bool stop_further_inputs = false;
+  for (int i = 0; i < input.events().size(); ++i) {
+    if (stop_further_inputs) {
+      break;
+    }
+    const auto& event = input.events(i);
+    ENVOY_LOG_MISC(debug, "Processing event: {}", event.DebugString());
+    // If we're disconnected, we fail out.
+    if (!tcp_client->connected()) {
+      ENVOY_LOG_MISC(debug, "Disconnected, no further event processing.");
+      break;
+    }
+    switch (event.event_selector_case()) {
+    case test::integration::Event::kDownstreamSendEvent: {
+      auto downstream_write_func = [&](const Http2Frame& h2_frame) -> void {
+        tcp_client->write(std::string(h2_frame), false, false);
+      };
+      // Start H2 session - send hello string
+      tcp_client->write(Http2Frame::Preamble, false, false);
```

Comment:
Is the H2 preamble sent on every downstream send event?

Reply:
Good catch, modified to send before the first downstream send event.